src/lang/lang_values.ml

```diff
@@ -460,18 +481,104 @@ module V = struct
             (print_term x)
       | Fun _ | FFI _ -> "<fun>"
 
-  let map_env f (env : env) = List.map (fun (s, v) -> (s, f v)) env
+  (** Find a method in a value. *)
+  let rec invoke x l =
+    match x.value with
+      | Meth (l', y, _) when l' = l -> y
+      | Meth (_, _, x) -> invoke x l
+      | _ -> failwith ("Could not find method " ^ l ^ " of " ^ print_value x)
 
-  let map_lazy_env f (env : lazy_env) =
-    List.map (fun (s, v) -> (s, Lazy.from_fun (fun () -> f Lazy.force v))) env
+  (** Perform a sequence of invokes: invokes x [l1;l2;l3;...] is x.l1.l2.l3... *)
+  let rec invokes x = function l :: ll -> invokes (invoke x l) ll | [] -> x
+
+  let rec demeth v = match v.value with Meth (_, _, v) -> demeth v | _ -> v
 end
 
 (** {2 Built-in values and toplevel definitions} *)
 
 let builtins : (((int * T.constraints) list * T.t) * V.value) Plug.plug =
   Plug.create ~duplicates:false ~doc:"scripting values" "scripting values"
 
-(* {2 Type checking/inference} *)
+(* Environment for builtins. *)
+let builtins_env : (string * (T.scheme * V.value)) list ref = ref []
+let default_environment () = !builtins_env
+
+let default_typing_environment () =
+  List.map (fun (x, (t, _)) -> (x, t)) !builtins_env
+
+let add_builtin ?(override = false) ?(register = true) ?doc name ((g, t), v) =
+  if register then builtins#register ?doc (String.concat "." name) ((g, t), v);
+  match name with
+    | [name] ->
+        (* Don't allow overriding builtins. *)
+        if (not override) && List.mem_assoc name !builtins_env then
+          failwith ("Trying to override builtin " ^ name);
+        builtins_env := (name, ((g, t), v)) :: !builtins_env
+    | x :: ll ->
+        let (g0, t0), xv =
+          try List.assoc x !builtins_env
+          with Not_found -> failwith ("Could not find builtin variable " ^ x)
+        in
+        (* x.l1.l2.l3 = v means
+           x = (x where l1 = (x.l1 where l2 = (x.l1.l2 where l3 = v)))
+        *)
+        (*
+let _ =
+  let x = "x" in
+  let v = "v" in
+  let rec aux prefix = function
+    | l::ll -> Printf.sprintf "%s.%s where (%s = %s)" x (String.concat "." (List.rev prefix)) l (aux (l::prefix) ll)
+    | [] -> v
+  in
+  let ll = ["l1"; "l2"; "l3"] in
+  Printf.sprintf "%s = %s" x (aux [] ll)
+    *)
```

Comment:
```suggestion
```

Reply:
It took me quite some time to get this one correctly, I'd rather leave this intermediate code for now just in case we need to make some changes on it...

Reply:
What's the purpose of this code then? How about making it a OCaml test?